File: backend/app/services/modules/base_module.py
from abc import ABC, abstractmethod
from typing import Dict, Any, List

class Question(ABC):
    question_string: str
    correct_answer: Any

class BaseQuizModule(ABC):
    slug: str
    title: str
    description: str
    
    @abstractmethod
    def generate_questions(self, config: Dict[str, Any]) -> List[Question]:
        """Generate a series of questions based on a dynamic config."""
        pass

    @abstractmethod
    def evaluate_answer(self, question: Question, user_answer: Any) -> bool:
        """Determines if the answer provided is acceptable."""
        pass

File: frontend/src/lib/quiz/base_module.ts
export interface Question {
  question_string: string;
  correct_answer: any;
}

export abstract class BaseQuizModule {
  abstract slug: string;
  abstract title: string;
  abstract description: string;

  /**
   * Given a dict/record of custom configs, return a list of questions.
   * Optionally takes a list of performance-based "weak spots" to influence sampling.
   */
  abstract generateQuestions(config: Record<string, any>, weakSpots?: string[]): Question[];

  /**
   * Evaluate a single answer for a question
   */
  abstract evaluateAnswer(question: Question, userAnswer: any): boolean;
}
